File: reference/policy.py
"""UZA Bulk commercial policy. Every number a founder should be able to change
without a developer lives here, not scattered through workflow code."""
from dataclasses import dataclass, field
from typing import Optional

# --- payment ------------------------------------------------------------------
# Collection must never sit behind payment to the factory.
PAYMENT_SCHEDULES = {
    "new":         [("confirmation", 0.50), ("pre_loading", 0.50)],
    "established": [("confirmation", 0.30), ("pre_loading", 0.40), ("pre_release", 0.30)],
}
MIN_DEPOSIT = 0.30              # never below this on manufacturing
ESTABLISHED_AFTER_ORDERS = 3    # completed, no late payment

# --- volumetrics --------------------------------------------------------------
CBM_TOLERANCE = 0.05            # declared vs measured: normal factory sloppiness
CBM_HARD_STOP = 0.10            # beyond this, loading stops until someone decides
FREIGHT_CONTINGENCY = 0.09      # built into every quotation

# --- commission ---------------------------------------------------------------
COMMISSION_RATE = 0.02          # on confirmed orders (deposit verified)
LEAD_DECAY_RATE = 0.01          # subsequent orders, same client, within 12 months
LEAD_OWNERSHIP_MONTHS = 12

# --- cost ladder --------------------------------------------------------------
# Order matters: each rung accumulates into the next incoterm.
LADDER = [
    ("exw",         "EXW"),
    ("inland_cn",   "FOB"),
    ("export_docs", "FOB"),
    ("origin_thc",  "FOB"),
    ("ocean",       "CIF"),
    ("insurance",   "CIF"),
    ("dest_charges", "DAP"),
    ("duty_vat",    "DAP"),
    ("inland_dest", "DAP"),
]
INCOTERMS = ["EXW", "FOB", "CIF", "DAP"]


@dataclass
class CostLine:
    est: float = 0.0
    act: Optional[float] = None

    @property
    def value(self):
        return self.act if self.act is not None else self.est

    @property
    def is_actual(self):
        return self.act is not None


@dataclass
class CostLadder:
    """Estimated at quotation, actuals filled in as they land."""
    lines: dict = field(default_factory=lambda: {k: CostLine() for k, _ in LADDER})
    basis: str = "EXW"          # EXW when inland is separable, FOB when supplier won't split
    inland_separable: bool = True

    def set_est(self, **kw):
        for k, v in kw.items():
            self.lines[k].est = v
        return self

    def set_act(self, **kw):
        for k, v in kw.items():
            self.lines[k].act = v
        return self

    def at(self, incoterm):
        """Cumulative cost up to and including the given incoterm."""
        idx = INCOTERMS.index(incoterm)
        return round(sum(self.lines[k].value for k, rung in LADDER
                         if INCOTERMS.index(rung) <= idx), 2)

    def margin(self, sell_price, incoterm):
        c = self.at(incoterm)
        return round((sell_price - c) / sell_price, 4) if sell_price else 0.0

    def actuals_landed(self):
        return all(self.lines[k].is_actual for k, _ in LADDER)


def revenue_ton(cbm, kg):
    """Forwarders bill on the greater of volume or weight."""
    return round(max(cbm, kg / 1000.0), 3)


def schedule_for(completed_orders):
    tier = "established" if completed_orders >= ESTABLISHED_AFTER_ORDERS else "new"
    return tier, PAYMENT_SCHEDULES[tier]

File: packages/contracts/src/policy.ts
/**
 * Commercial policy. Every number a founder should be able to change without a
 * developer lives here. Modules must read these constants, never inline them.
 */

// ---------- payment ----------
export type InstallmentTrigger = 'confirmation' | 'pre_loading' | 'pre_release';
export type ClientTier = 'new' | 'established';

export const PAYMENT_SCHEDULES: Record<ClientTier, ReadonlyArray<readonly [InstallmentTrigger, number]>> = {
  new:         [['confirmation', 0.50], ['pre_loading', 0.50]],
  established: [['confirmation', 0.30], ['pre_loading', 0.40], ['pre_release', 0.30]],
};

export const MIN_DEPOSIT = 0.30;
export const ESTABLISHED_AFTER_ORDERS = 3;

export const scheduleFor = (completedOrders: number) => {
  const tier: ClientTier = completedOrders >= ESTABLISHED_AFTER_ORDERS ? 'established' : 'new';
  return { tier, schedule: PAYMENT_SCHEDULES[tier] };
};

// ---------- volumetrics ----------
export const CBM_TOLERANCE = 0.05;   // declared vs measured, normal factory drift
export const CBM_HARD_STOP = 0.10;   // beyond this, loading stops until a human decides
export const FREIGHT_CONTINGENCY = 0.09;
export const BILLING_CLAIM_THRESHOLD = 0.02; // billed over measured before we claim
/** Revenue-ton capacity of a container, for utilisation reporting only (never a gate).
 * Ties to the open founder decision on container utilisation (CLAUDE.md §7). */
export const CONTAINER_RT_CAPACITY = 28.0;

export type VarianceDecision = 'client_pays' | 'uza_absorbs' | 'reduce_qty';

// ---------- commission ----------
export const COMMISSION_RATE = 0.02;       // on confirmed orders (deposit verified)
export const LEAD_DECAY_RATE = 0.01;       // repeat orders, same client
export const LEAD_OWNERSHIP_MONTHS = 12;
export const CLAIMS_WINDOW_DAYS = 14;

// ---------- quotation pricing ----------
/**
 * The negotiation floor and walk-away ceiling, derived from the supplier unit cost.
 * TARGET is the price a negotiator aims for; below WALKAWAY UZA declines the deal.
 * Founder-tunable commercial policy — never inline these in a pricing routine.
 */
export const TARGET_PRICE_FACTOR = 0.92;   // negotiation floor vs supplier unit cost
export const WALKAWAY_FACTOR = 1.05;       // walk-away ceiling vs supplier unit cost

// ---------- quality: inspection grading ----------
/**
 * How an inspection's defect counts grade into a result. A critical defect always fails
 * and blocks release (no override); more than the major threshold is only conditional.
 * Founder-tunable quality policy — quality grades against it, the release gate and the
 * web inspection screen read the same boundary.
 */
export const INSPECTION_THRESHOLDS = {
  /** critical > this ⇒ fail. A critical defect always fails, no override. */
  CRITICAL_FAIL_AT: 0,
  /** major > this ⇒ conditional (when not already failing on a critical). */
  MAJOR_CONDITIONAL_AT: 2,
} as const;

/**
 * The single definition of the grading rule, so a publisher and the release gate never
 * re-derive "> 2 major = conditional" and drift. Return type mirrors Inspection.result.
 */
export const gradeInspection = (
  critical: number,
  major: number,
): 'pass' | 'conditional' | 'fail' => {
  if (critical > INSPECTION_THRESHOLDS.CRITICAL_FAIL_AT) return 'fail';
  if (major > INSPECTION_THRESHOLDS.MAJOR_CONDITIONAL_AT) return 'conditional';
  return 'pass';
};

// ---------- sourcing: supplier score ----------
/**
 * How hard a factory's score is lowered for declared-vs-measured volumetric variance:
 *   delta = -min(abs(variance) * VARIANCE_MULTIPLIER, VARIANCE_CAP)
 * applied only when the receipt flagged a discrepancy. Founder-tunable — finance and
 * supplier analytics read the resulting score series; warehouse produces the variance.
 */
export const SUPPLIER_SCORE = {
  VARIANCE_MULTIPLIER: 10,   // score points per unit of |variance|
  VARIANCE_CAP: 2.0,         // max single-event penalty
} as const;

// ---------- cost ladder ----------
export const INCOTERMS = ['EXW', 'FOB', 'CIF', 'DAP'] as const;
export type Incoterm = (typeof INCOTERMS)[number];

/** Each rung accumulates into the incoterm it is listed against. */
export const LADDER: ReadonlyArray<readonly [CostRung, Incoterm]> = [
  ['exw', 'EXW'],
  ['inlandCn', 'FOB'], ['exportDocs', 'FOB'], ['originThc', 'FOB'],
  ['ocean', 'CIF'], ['insurance', 'CIF'],
  ['destCharges', 'DAP'], ['dutyVat', 'DAP'], ['inlandDest', 'DAP'],
];

export type CostRung =
  | 'exw' | 'inlandCn' | 'exportDocs' | 'originThc'
  | 'ocean' | 'insurance' | 'destCharges' | 'dutyVat' | 'inlandDest';

/**
 * Split a total into installments whose parts sum EXACTLY to the total.
 *
 * Naive rounding of 30/40/30 leaves cents unaccounted for, and the last
 * installment then never marks itself paid. The remainder goes to the final
 * installment deliberately.
 */
export const splitInstallments = (
  totalMinor: number,
  schedule: ReadonlyArray<readonly [InstallmentTrigger, number]>,
): ReadonlyArray<{ trigger: InstallmentTrigger; pct: number; amountMinor: number }> => {
  const parts = schedule.map(([trigger, p]) => ({
    trigger, pct: p, amountMinor: Math.floor(totalMinor * p),
  }));
  const allocated = parts.reduce((a, p) => a + p.amountMinor, 0);
  const last = parts[parts.length - 1];
  if (last) last.amountMinor += totalMinor - allocated;
  return parts;
};
